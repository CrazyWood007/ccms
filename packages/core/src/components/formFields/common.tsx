import React from 'react'
import { ColumnsConfig, ParamConfig } from '../../interface'

import { FieldConfigs as getFieldConfigs } from './'
import ParamHelper from '../../util/param'
import { updateCommonPrefixItem } from '../../util/value'
import { ConditionConfig } from '../../util/condition'
import { StatementConfig } from '../../util/statement'
import { isEqual, get } from 'lodash'

/**
 * 表单项基类配置文件格式定义
 * - field:    表单项字段名
 * - type:     表单项类型
 * - * text:     短文本类型
 * - * form:     子表单类型
 * - label:    表单项名称
 * - required: 表单项必填
 * - readonly: 表单项只读
 * - disabled: 表单项不可编辑
 * - default:  表单项默认值  // 改为defaultValue
 * - - type:     默认值类型
 * - - * static:   固定值
 * - - * data:     上一步骤数据
 * - - * query:    页面GET方法传参
 * - - * hash:     页面HASH传参
 * - - * interface: 接口入参获取
 * - - value:    默认值（static类型使用）
 * - - field:    字段名（data/query/hash类型使用）（hash类型选填）
 */
export interface FieldConfig {
  field: string
  label: string
  required?: boolean
  readonly?: boolean
  disabled?: boolean
  display?: 'none'
  defaultValue?: ParamConfig,
  condition?: ConditionConfig
  extra?: StatementConfig
  columns?: ColumnsConfig
  styles?: object
}

/**
 * 表单项配置文件格式定义 - 枚举
 */
export type FieldConfigs = getFieldConfigs

/**
 * 表单项子类需实现的方法
 * - reset:    表单项重置当前值
 * - set:      表单项设置当前值
 * - get:      表单项获取当前值
 * - validate: 表单项的值校验方法
 */
export interface IField<T> {
  reset: () => Promise<T>
  set: (value: T) => Promise<void>
  get: () => Promise<T>
  validate: (value: T) => Promise<true | FieldError[]>
  fieldFormat: () => Promise<{}>
}

/**
 * 表单项子类需要的入参
 * - ref:
 * - formLayout:
 * - value:
 * - data:
 * - step:
 * - config:
 * - onChange:
 */
export interface FieldProps<C extends FieldConfig, T> {
  // 挂载事件
  ref: (instance: Field<C, {}, any> | null) => void
  // 挂载引用
  form: React.ReactNode
  formLayout: 'horizontal' | 'vertical' | 'inline'
  value: T,
  record: { [field: string]: any },
  data: any[],
  config: C
  // TODO 待删除
  onChange: (value: T) => Promise<void>
  // 事件：设置值  noPathCombination：为true时不做路径拼接
  onValueSet: (path: string, value: T, validation: true | FieldError[], options?: { noPathCombination?: boolean }) => Promise<void>
  // 事件：置空值
  onValueUnset: (path: string, validation: true | FieldError[], options?: { noPathCombination?: boolean }) => Promise<void>
  // 事件：修改值 - 列表 - 追加
  onValueListAppend: (path: string, value: any, validation: true | FieldError[], options?: { noPathCombination?: boolean }) => Promise<void>
  // 事件：修改值 - 列表 - 删除
  onValueListSplice: (path: string, index: number, count: number, validation: true | FieldError[], options?: { noPathCombination?: boolean }) => Promise<void>
  // 事件：修改值 - 列表 - 修改顺序
  onValueListSort: (path: string, index: number, sortType: 'up' | 'down', validation: true | FieldError[], options?: { noPathCombination?: boolean }) => Promise<void>
  baseRoute: string,
  containerPath: string, // 容器组件所在路径以字段拼接展示  1.3.0新增
  onReportFields?: (field: string) => Promise<void> // 向父组件上报依赖字段  1.3.0新增
  step: { [field: string]: any } // 传递formValue
  loadDomain: (domain: string) => Promise<string>
}

/**
 * 表单项配置接口获取数据需要的入参
* - url: 请求地址
* - method: 请求类型
* - withCredentials?: 跨域是否提供凭据信息
* - response: 返回值
* - format?: 格式化返回值
* - responseArrayKey?: format === 'array' 时配置 key 值
* - responseArrayValue?: format === 'array' 时配置 value 值
 */
export interface FieldInterface {
  interface?: {
    url: string
    method: 'GET' | 'POST' | 'get' | 'post'
    withCredentials?: boolean
    response: string
    format?: 'array' | 'key'
    responseArrayKey?: string
    responseArrayValue?: string
  }
}

/**
 * 表单项基类
 * - C: 表单项的配置文件类型
 * - E: 表单项的渲染方法入参
 * - T: 表单项的值类型
 * - S: 表单项的扩展状态
 */
export class Field<C extends FieldConfig, E, T, S = {}> extends React.Component<FieldProps<C, T>, S> implements IField<T> {
  dependentFields: string[] = [] // 组件param依赖字段存放数组  1.3.0新增
  static defaultProps = {
    config: {}
  };

  /**
   * 获取默认值
   */
  defaultValue : () => Promise<any> = async () => {
    const {
      config
    } = this.props
    if (config.defaultValue !== undefined) {
      return ParamHelper(config.defaultValue, { record: this.props.record, data: this.props.data, step: this.props.step }, this)
    }

    return undefined
  }

  reset: () => Promise<T> = async () => {
    return this.defaultValue()
  };

  set: (value: any) => Promise<any> = async (value) => {
    return value
  };

  get: () => Promise<T> = async () => {
    return this.props.value
  }

  validate: (value: T) => Promise<true | FieldError[]> = async () => {
    return true
  };

  fieldFormat: () => Promise<{}> = async () => {
    return {}
  }

  didMount: () => Promise<void> = async () => { }

  /**
   * 上报param依赖字段名称
   * @param field
   */
  handleReportFields: (field: string) => void = async (field) => {
    const update: string[] | boolean = updateCommonPrefixItem(this.dependentFields, field)
    if (typeof update === 'boolean') return
    this.dependentFields = update
    this.props.onReportFields && await this.props.onReportFields(field)
  }

  renderComponent = (props: E) => {
    return <React.Fragment>
      当前UI库未实现该表单类型
    </React.Fragment>
  }

  shouldComponentUpdate (nextProps: FieldProps<C, T>, nextState: S) {
    // console.log('nextProps', nextProps, this.props, nextProps.value == this.props.value);

    const dependentFieldsArr = this.dependentFields
    // console.log('dependentFieldsArr',dependentFieldsArr);
    let dependentIsChange = false
    if (dependentFieldsArr && dependentFieldsArr.length) {
      for (let i = dependentFieldsArr.length; i >= 0; i--) {
        const nextDependentField = get(nextProps.step, dependentFieldsArr[i])
        const currentDependentField = get(this.props.step, dependentFieldsArr[i])

        if ((nextDependentField || currentDependentField) && nextDependentField !== currentDependentField) {
          dependentIsChange = true
          break
        }
      }
    }

    /**
     * data提交前不变, 去掉这项的比较
     * record也不比较，需要比较的话就在dependentFieldsArr取出record绝对路径
     * */
    if (!dependentIsChange && isEqual(this.state, nextState) && nextProps.value === this.props.value && this.props.config === nextProps.config) {
      // console.log('no update' );
      return false
    }
    return true
  }

  render = () => {
    return (<React.Fragment>
      当前UI库未实现该表单类型
    </React.Fragment>)
  }
}

export interface DisplayProps<C extends FieldConfig, T> {
  value: T,
  record: { [field: string]: any },
  data: any[],
  step: { [field: string]: any }
  config: C,
  // 事件：设置值
  onValueSet: (path: string, value: T, options?: { noPathCombination?: boolean }) => Promise<void>
  // 事件：置空值
  onValueUnset: (path: string, options?: { noPathCombination?: boolean }) => Promise<void>
  // 事件：修改值 - 列表 - 追加
  onValueListAppend: (path: string, value: any, options?: { noPathCombination?: boolean }) => Promise<void>
  // 事件：修改值 - 列表 - 删除
  onValueListSplice: (path: string, index: number, count: number, options?: { noPathCombination?: boolean }) => Promise<void>
  baseRoute: string,
  loadDomain: (domain: string) => Promise<string>
}

export class Display<C extends FieldConfig, E, T, S = {}> extends React.Component<DisplayProps<C, T>, S> {
  defaultValue = async () => {
    const {
      config
    } = this.props
    if (config.defaultValue !== undefined) {
      return ParamHelper(config.defaultValue, { record: this.props.record, data: this.props.data, step: this.props.step })
    }

    return undefined
  }

  reset: () => Promise<T> = async () => {
    return this.defaultValue()
  };

  set: (value: any) => Promise<any> = async (value) => {
    return value
  };

  get: () => Promise<T> = async () => {
    return this.props.value
  }

  renderComponent = (props: E) => {
    return <React.Fragment>
      当前UI库未实现该表单类型
    </React.Fragment>
  }

  didMount: () => Promise<void> = async () => { }

  render = () => {
    return (<React.Fragment>
      当前UI库未实现该表单类型
    </React.Fragment>)
  }
}

export class FieldError {
  message: string
  constructor (message: string) {
    this.message = message
  }
}
