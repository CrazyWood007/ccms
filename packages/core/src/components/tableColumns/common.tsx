import React from 'react'
import { componentType } from '.'

export interface ColumnConfig extends componentType {
  field: string
  label: string,
  defaultValue?: string,
  style?: {
    color?: string
    fontSize?: number
    prefix?: string
    postfix?: string
    customStyle?: object
  }
}

/**
 * 表格项项子类需实现的方法
 * - get:      表格项获取展示值
 */
export interface IColumn<T> {
  getValue: () => Promise<T>
}

export interface ColumnProps<T> {
  ref: (instance: Column<T, any> | null) => void
  record: { [field: string]: any }
  value: any
  data: any[],
  step: number,
  config: T
}

interface ColumnState {
}

export default class Column<T, E> extends React.Component<ColumnProps<T>, ColumnState> implements IColumn<T> {
  constructor (props: ColumnProps<T>) {
    super(props)
    this.state = {}
  }
  
  static defaultProps = {
    config: {}
  }

  getValue= ()=>{
    return this.props.value
  }

  renderComponent = (props: E) => {
    return <React.Fragment>
      您当前使用的UI版本没有实现Column组件。
    </React.Fragment>
  }

  render = () => {
    return (<></>)
  }
}
