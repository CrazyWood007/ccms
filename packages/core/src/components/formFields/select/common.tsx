import { ReactNode } from 'react'
import EnumerationHelper, { EnumerationOptionsConfig, InterfaceEnumerationOptionsKVConfig, InterfaceEnumerationOptionsListConfig } from '../../../util/enumeration'
import InterfaceHelper from '../../../util/interface'
import { Field, FieldConfig, FieldProps, IField, Display, DisplayProps } from '../common'

export interface SelectFieldConfig extends FieldConfig {
  options?: EnumerationOptionsConfig
  defaultSelect?: boolean | number
}

export interface ISelectFieldOption {
  value: string | number | boolean,
  label: ReactNode,
  children?: Array<ISelectFieldOption>
}

interface SelectSingleFieldState {
  options: Array<{
    value: string | number | boolean
    label: string
    [extra: string]: any
  }>
}

export default class SelectField<C extends SelectFieldConfig, E, T> extends Field<C, E, T, SelectSingleFieldState> implements IField<T> {
  interfaceHelper = new InterfaceHelper()

  constructor(props: FieldProps<C, T>) {
    super(props)

    this.state = {
      options: []
    }
  }

  options = (
    config: EnumerationOptionsConfig | undefined
  ) => {
    if (config) {
      EnumerationHelper.options(
        config,
        (config, source) => this.interfaceHelper.request(config, source, { record: this.props.record, data: this.props.data, step: this.props.step }, { loadDomain: this.props.loadDomain }, this),
        { record: this.props.record, data: this.props.data, step: this.props.step }
      ).then((options) => {
        if (JSON.stringify(this.state.options) !== JSON.stringify(options)) {
          this.setState({
            options
          })
        }
      })

      return this.state.options
    }
    return []
  }
}

export class SelectDisplay<C extends SelectFieldConfig, E, T> extends Display<C, E, T, SelectSingleFieldState> {
  interfaceHelper = new InterfaceHelper()

  constructor(props: DisplayProps<C, T>) {
    super(props)

    this.state = {
      options: []
    }
  }

  options = (
    config: EnumerationOptionsConfig | undefined
  ) => {
    if (config) {
      EnumerationHelper.options(
        config,
        (config, source) => this.interfaceHelper.request(config, source, { record: this.props.record, data: this.props.data, step: this.props.step }, { loadDomain: this.props.loadDomain }),
        { record: this.props.record, data: this.props.data, step: this.props.step }
      ).then((options) => {
        if (JSON.stringify(this.state.options) !== JSON.stringify(options)) {
          this.setState({
            options
          })
        }
      })

      return this.state.options
    }
    return []
  }
}
