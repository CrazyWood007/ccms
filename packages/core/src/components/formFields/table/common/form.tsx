import marked from 'marked'
import React from 'react'
import getALLComponents, { FieldConfigs } from '../..'
import { ColumnsConfig } from '../../../../interface'
import { CCMSConfig, PageListItem } from '../../../../main'
import { IFormItem } from '../../../../steps/form'
import ConditionHelper from '../../../../util/condition'
import { push, set, sort, splice } from '../../../../util/produce'
import StatementHelper from '../../../../util/statement'
import { getBoolean, getValue, updateCommonPrefixItem } from '../../../../util/value'
import { Field, FieldError } from '../../common'

export interface TableFieldFormProps {
  form: React.ReactNode
  formLayout: 'horizontal' | 'vertical' | 'inline'
  config: { fields: FieldConfigs[]; childColumns?: ColumnsConfig }
  data: { [field: string]: unknown }
  datas: {
    record?: object | undefined
    data: object[]
    step: {
      [field: string]: unknown
    }
    containerPath: string
  }
  checkPageAuth: (pageID: unknown) => Promise<boolean>
  loadPageURL: (pageID: unknown) => Promise<string>
  loadPageFrameURL: (pageID: unknown) => Promise<string>
  loadPageConfig: (pageID: unknown) => Promise<CCMSConfig>
  loadPageList: () => Promise<Array<PageListItem>>
  baseRoute: string
  loadDomain: (domain: string) => Promise<string>
  containerPath: string // 容器组件所在路径以字段拼接展示  1.3.0新增
  onReportFields?: (field: string) => Promise<void> // 向父组件上报依赖字段  1.3.0新增

  children: (onClick: () => void) => JSX.Element | JSX.Element[]
  onSubmit: (value: { [key: string]: unknown }) => void
}

interface TableFieldFormState {
  visible: boolean
  formData: { status: 'normal' | 'error' | 'loading'; message?: string; name: string }[]
}

export interface ITableFieldFormModal {
  title: string
  content: React.ReactNode
  onOk: () => void
  onClose: () => void
}

export default class TableFieldForm extends React.Component<TableFieldFormProps, TableFieldFormState> {
  dependentFields: string[] = []

  formFieldsMounted: Array<boolean> = []

  formFields: Array<Field<FieldConfigs, unknown, unknown> | null> = []

  formValue: { [key: string]: unknown } = {}

  formData: { status: 'normal' | 'error' | 'loading'; message?: string; name: string }[] = []

  constructor(props) {
    super(props)

    this.state = {
      visible: false,
      formData: []
    }
  }

  handleSubLabelContent(config) {
    const { datas } = this.props
    if (config?.subLabelConfig?.enable) {
      const content = StatementHelper(
        {
          statement: config.subLabelConfig?.content?.statement || '',
          params: config.subLabelConfig?.content?.params || []
        },
        { ...datas, record: this.formValue }
      ).replace(/(^\s*)|(\s*$)/g, '')
      const mode = config.subLabelConfig?.mode
      switch (mode) {
        case 'markdown':
          // eslint-disable-next-line react/no-danger
          return <div dangerouslySetInnerHTML={{ __html: marked(content) }} />
        case 'html':
          // eslint-disable-next-line react/no-danger
          return <div style={{ whiteSpace: 'pre-wrap' }} dangerouslySetInnerHTML={{ __html: content }} />
        default:
          return <div style={{ whiteSpace: 'pre-wrap' }}>{content}</div>
      }
    }
    return undefined
  }

  getALLComponents = (type: string): typeof Field => getALLComponents[type]

  handleReportFields: (field: string) => void = async (field) => {
    const { onReportFields } = this.props
    const update: string[] | boolean = updateCommonPrefixItem(this.dependentFields, field)
    if (typeof update === 'boolean') return
    this.dependentFields = update
    onReportFields && (await onReportFields(field))
  }

  handleMount = async (formFieldIndex: number) => {
    if (this.formFieldsMounted[formFieldIndex]) {
      return true
    }
    this.formFieldsMounted = set(this.formFieldsMounted, `[${formFieldIndex}]`, true)

    if (this.formFields[formFieldIndex]) {
      const formField = this.formFields[formFieldIndex]
      if (formField) {
        const { config } = this.props
        const formFieldConfig = (config.fields || [])[formFieldIndex]
        let value = getValue(this.formValue, formFieldConfig.field)
        const source = value
        if (formFieldConfig.defaultValue && value === undefined) {
          value = await formField.reset()
        }
        value = await formField.set(value)
        if (source !== value) {
          set(this.formValue, formFieldConfig.field, value)
        }

        if (value !== undefined) {
          const validation = await formField.validate(value)
          this.handleValueCallback(formFieldIndex, validation)
        }
        await formField.didMount()
      }
    }
  }

  handleValueCallback = async (formFieldIndex: number, validation: true | FieldError[]) => {
    let { formData } = this.state
    if (validation === true) {
      formData = set(formData, `[${formFieldIndex}]`, { status: 'normal' })
    } else {
      formData = set(formData, `[${formFieldIndex}]`, { status: 'error', message: validation[0].message })
    }

    await this.setState({
      formData
    })
  }

  handleValueSet = async (
    formFieldIndex: number,
    path: string,
    value: unknown,
    validation: true | FieldError[],
    options?: { noPathCombination?: boolean }
  ) => {
    const { config } = this.props
    const formFieldConfig = (config.fields || [])[formFieldIndex]
    if (formFieldConfig) {
      let fullPath = ''
      if (options && options.noPathCombination) {
        fullPath = path
      } else if (formFieldConfig.field === '' || path === '') {
        fullPath = `${formFieldConfig.field}${path}`
      } else {
        fullPath = `${formFieldConfig.field}.${path}`
      }

      this.formValue = set(this.formValue, fullPath, value)
      this.handleValueCallback(formFieldIndex, validation)
    }
  }

  handleValueUnset = async (
    formFieldIndex: number,
    path: string,
    validation: true | FieldError[],
    options?: { noPathCombination?: boolean }
  ) => {
    const { config } = this.props
    const formFieldConfig = (config.fields || [])[formFieldIndex]
    if (formFieldConfig) {
      let fullPath = ''
      if (options && options.noPathCombination) {
        fullPath = path
      } else if (formFieldConfig.field === '' || path === '') {
        fullPath = `${formFieldConfig.field}${path}`
      } else {
        fullPath = `${formFieldConfig.field}.${path}`
      }

      this.formValue = set(this.formValue, fullPath)
      this.handleValueCallback(formFieldIndex, validation)
    }
  }

  handleValueListAppend = async (
    formFieldIndex: number,
    path: string,
    value: unknown,
    validation: true | FieldError[],
    options?: { noPathCombination?: boolean }
  ) => {
    const { config } = this.props
    const formFieldConfig = (config.fields || [])[formFieldIndex]
    if (formFieldConfig) {
      let fullPath = ''
      if (options && options.noPathCombination) {
        fullPath = path
      } else if (formFieldConfig.field === '' || path === '') {
        fullPath = `${formFieldConfig.field}${path}`
      } else {
        fullPath = `${formFieldConfig.field}.${path}`
      }

      this.formValue = push(this.formValue, fullPath, value)
      this.handleValueCallback(formFieldIndex, validation)
    }
  }

  handleValueListSplice = async (
    formFieldIndex: number,
    path: string,
    index: number,
    count: number,
    validation: true | FieldError[],
    options?: { noPathCombination?: boolean }
  ) => {
    const { config } = this.props
    const formFieldConfig = (config.fields || [])[formFieldIndex]
    if (formFieldConfig) {
      let fullPath = ''
      if (options && options.noPathCombination) {
        fullPath = path
      } else if (formFieldConfig.field === '' || path === '') {
        fullPath = `${formFieldConfig.field}${path}`
      } else {
        fullPath = `${formFieldConfig.field}.${path}`
      }

      this.formValue = splice(this.formValue, fullPath, index, count)
      this.handleValueCallback(formFieldIndex, validation)
    }
  }

  handleValueListSort = async (
    formFieldIndex: number,
    path: string,
    index: number,
    sortType: 'up' | 'down' | 'top' | 'bottom',
    validation: true | FieldError[],
    options?: { noPathCombination?: boolean }
  ) => {
    const { config } = this.props
    const formFieldConfig = (config.fields || [])[formFieldIndex]
    if (formFieldConfig) {
      let fullPath = ''
      if (options && options.noPathCombination) {
        fullPath = path
      } else if (formFieldConfig.field === '' || path === '') {
        fullPath = `${formFieldConfig.field}${path}`
      } else {
        fullPath = `${formFieldConfig.field}.${path}`
      }

      this.formValue = sort(this.formValue, fullPath, index, sortType)
      this.handleValueCallback(formFieldIndex, validation)
    }
  }

  renderModal: (props: ITableFieldFormModal) => JSX.Element = () => {
    return <>您当前使用的UI版本没有实现OpertionHelper组件。</>
  }

  renderItemComponent: (props: IFormItem) => JSX.Element = () => {
    return <>您当前使用的UI版本没有实现FormItem组件。</>
  }

  render() {
    const {
      form,
      formLayout,
      config,
      data,
      datas,
      children,
      checkPageAuth,
      loadPageURL,
      loadPageFrameURL,
      loadPageConfig,
      loadPageList,
      baseRoute,
      loadDomain,
      containerPath
    } = this.props

    const { visible, formData } = this.state

    return (
      <>
        {children(() => {
          this.formValue = data
          this.setState({ visible: true })
        })}
        {visible &&
          this.renderModal({
            title: 'title',
            content: (config.fields || []).map((formFieldConfig, formFieldIndex) => {
              if (!ConditionHelper(formFieldConfig.condition, { ...datas, record: this.formValue })) {
                this.formFieldsMounted = set(this.formFieldsMounted, `[${formFieldIndex}]`, false)
                this.formFields && (this.formFields[formFieldIndex] = null)
                return null
              }
              let hidden = true
              let display = true

              if (formFieldConfig.type === 'hidden') {
                hidden = true
                display = false
              }

              if (formFieldConfig.display === 'none') {
                hidden = true
                display = false
              }

              // 隐藏项同时打标录入数据并清空填写项
              if (!hidden) {
                this.formData = set(this.formData, `[${formFieldIndex}]`, {
                  status: 'normal',
                  name: formFieldConfig.label
                })
              }

              const FormField = this.getALLComponents(formFieldConfig.type) || Field

              let status = formData[formFieldIndex]?.status || 'normal'

              if (['group', 'import_subform', 'object', 'tabs', 'form'].some((type) => type === formFieldConfig.type)) {
                status = 'normal'
              }

              const renderData = {
                key: formFieldIndex,
                label: formFieldConfig.label,
                subLabel: this.handleSubLabelContent(formFieldConfig),
                styles: formFieldConfig.styles,
                status,
                message: (formData[formFieldIndex] || {}).message || '',
                extra: StatementHelper(formFieldConfig.extra, { ...datas, record: this.formValue }),
                required: getBoolean(formFieldConfig.required),
                layout: formLayout,
                visitable: display,
                fieldType: formFieldConfig.type,
                children: (
                  <FormField
                    key={formFieldIndex}
                    ref={(formField: Field<FieldConfigs, unknown, unknown> | null) => {
                      if (formField) {
                        this.formFields = set(this.formFields, `[${formFieldIndex}]`, formField)
                        this.handleMount(formFieldIndex)
                      }
                    }}
                    form={form}
                    formLayout={formLayout}
                    value={getValue(this.formValue, formFieldConfig.field)}
                    record={this.formValue}
                    data={datas.data}
                    step={datas.step}
                    config={formFieldConfig}
                    onChange={async () => {
                      /* 无逻辑 */
                    }}
                    onValueSet={async (path, valueSet, validation, options) =>
                      this.handleValueSet(formFieldIndex, path, valueSet, validation, options)
                    }
                    onValueUnset={async (path, validation, options) =>
                      this.handleValueUnset(formFieldIndex, path, validation, options)
                    }
                    onValueListAppend={async (path, valueAppend, validation, options) =>
                      this.handleValueListAppend(formFieldIndex, path, valueAppend, validation, options)
                    }
                    onValueListSplice={async (path, index, count, validation, options) =>
                      this.handleValueListSplice(formFieldIndex, path, index, count, validation, options)
                    }
                    onValueListSort={async (path, index, sortType, validation, options) =>
                      this.handleValueListSort(formFieldIndex, path, index, sortType, validation, options)
                    }
                    checkPageAuth={async (pageID) => checkPageAuth(pageID)}
                    loadPageURL={async (pageID) => loadPageURL(pageID)}
                    loadPageFrameURL={async (pageID) => loadPageFrameURL(pageID)}
                    loadPageConfig={async (pageID) => loadPageConfig(pageID)}
                    loadPageList={async () => loadPageList()}
                    baseRoute={baseRoute}
                    loadDomain={async (domain: string) => loadDomain(domain)}
                    containerPath={containerPath}
                    onReportFields={async (field: string) => this.handleReportFields(field)}
                  />
                )
              }
              // 渲染表单项容器
              return hidden ? this.renderItemComponent(renderData) : <React.Fragment key={formFieldIndex} />
            }),
            onOk: () => {
              const { onSubmit } = this.props
              onSubmit(this.formValue)
              this.setState({
                visible: false
              })
            },
            onClose: () => {
              this.setState({
                visible: false
              })
            }
          })}
      </>
    )
  }
}
