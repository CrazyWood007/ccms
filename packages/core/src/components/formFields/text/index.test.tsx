import React from 'react'
import { render, cleanup } from '@testing-library/react'
import renderer from 'react-test-renderer'
import TextField, { TextFieldConfig } from '.'
import { FieldError, FieldProps } from '../common'

// 默认入参
const defaultProps: FieldProps<TextFieldConfig, string> = {
  ref: async (ref: any) => { },
  formLayout: 'horizontal',
  value: '',
  data: [{}],
  step: 0,
  config: { field: 'jest', label: 'jest', type: 'text' },
  onChange: async () => { },
  record: {}
}

test('文本框默认值 - 未配置', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            const value = await ref.reset()
            expect(value).toEqual('')
          }
          cleanup()
          resolve(true)
        }}
      />
    )
  })
})

test('文本框默认值 - 静态值', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            const value = await ref.reset()
            expect(value).toEqual('default')
          }
          cleanup()
          resolve(true)
        }}
        config={{ field: 'jest', label: 'jest', type: 'text', default: { type: 'static', value: 'default' } }}
      />
    )
  })
})
test('文本框默认值 -数据值', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            const value = await ref.reset()
            expect(value).toEqual('')
          }
          cleanup()
          resolve(true)
        }}
        config={{ field: 'jest', label: 'jest', type: 'text', default: { type: 'data', value: 'default' } }}
      />
    )
  })
})
test('文本框默认值 -数据值 query', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            const value = await ref.reset()
            expect(value).toEqual('')
          }
          cleanup()
          resolve(true)
        }}
        config={{ field: 'jest', label: 'jest', type: 'text', default: { type: 'query', value: 'default' } }}
      />
    )
  })
})

test('文本框默认值 -数据值 hash', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            const value = await ref.reset()
            expect(value).toEqual('')
          }
          cleanup()
          resolve(true)
        }}
        config={{ field: 'jest', label: 'jest', type: 'text', default: { type: 'hash', value: 'default' } }}
      />
    )
  })
})
test('文本框默认值 - 接口值 -GET', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            const value = await ref.reset()
            expect(typeof value).toEqual('string')
          }
          cleanup()
          resolve(true)
        }}
        config={{
          field: 'jest',
          label: 'jest',
          type: 'text',
          default: {
            type: 'interface',
            api: {
              url: 'http://j-api.jd.com/mocker/data?p=263&v=POST&u=list.do',
              method: 'GET',
              contentType: 'json',
              withCredentials: true
            },
            apiResponse: 'result.0.datetime'
          }
        }}
      />
    )
  })
})
test('文本框默认值 - 接口值 -POST', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            const value = await ref.reset()
            expect(typeof value).toEqual('string')
          }
          cleanup()
          resolve(true)
        }}
        config={{
          field: 'jest',
          label: 'jest',
          type: 'text',
          default: {
            type: 'interface',
            api: {
              url: 'http://j-api.jd.com/mocker/data?p=263&v=POST&u=list.do',
              method: 'POST',
              contentType: 'json',
              withCredentials: true
            },
            apiResponse: 'result.0.datetime'
          }
        }}
      />
    )
  })
})
test('文本框默认值 - 接口值 -fial', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            const value = await ref.reset()
            expect(typeof value).toEqual('string')
          }
          cleanup()
          resolve(true)
        }}
        config={{
          field: 'jest',
          label: 'jest',
          type: 'text',
          default: {
            type: 'interface',
            api: {
              url: 'http://j-api.jd.com/mocker/data/',
              method: 'POST',
              contentType: 'json',
              withCredentials: true
            },
            apiResponse: 'result.0.datetime'
          }
        }}
      />
    )
  })
})
test('文本框解析', () => {
  return new Promise((resolve) => {
    render(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => {
          if (ref) {
            await ref.set('test')
          }
        }}
        onChange={async (value) => {
          expect(value).toEqual('test')
          cleanup()
          resolve(true)
        }}
      />
    )
  })
})

const validateTest = (message: string, config: TextFieldConfig, successValue: string, failValue: string, errorMessage?: string) => {
  test(message, () => {
    return new Promise((resolve) => {
      render(
        <TextField
          {...defaultProps}
          ref={async (ref: any) => {
            if (ref) {
              const success = await ref.validate(successValue)
              expect(success).toEqual(true)

              const fail = await ref.validate(failValue)
              expect(fail).toContainEqual(new FieldError(errorMessage || ''))

              cleanup()
              resolve(true)
            }
          }}
          config={config}
        />
      )
    })
  })
}

validateTest('文本框必填校验', Object.assign({ required: true }, defaultProps.config), '*', '', '不能为空')
validateTest('文本框类型校验-数字', Object.assign({ characterType: { enable: true, number: true } }, defaultProps.config), '1', '1aA中_-', '仅能输入数字')
validateTest('文本框类型校验-大写拉丁字母', Object.assign({ characterType: { enable: true, uppercase: true } }, defaultProps.config), 'A', '1aA中_-', '仅能输入大写拉丁字母')
validateTest('文本框类型校验-小写拉丁字母', Object.assign({ characterType: { enable: true, lowercase: true } }, defaultProps.config), 'a', '1aA中_-', '仅能输入小写拉丁字母')
validateTest('文本框类型校验-中日韩字符', Object.assign({ characterType: { enable: true, cjk: true } }, defaultProps.config), '中', '1aA中_-', '仅能输入中日韩字符')
validateTest('文本框类型校验-下划线', Object.assign({ characterType: { enable: true, underline: true } }, defaultProps.config), '_', '1aA中_-', '仅能输入下划线')
validateTest('文本框类型校验-中划线', Object.assign({ characterType: { enable: true, hyphen: true } }, defaultProps.config), '-', '1aA中_-', '仅能输入中划线')
validateTest('文本框最大长度校验', Object.assign({ maxLength: 5 }, defaultProps.config), '12345', '123456', '最长可输入5个字符。')
validateTest('文本框最小长度校验', Object.assign({ minLength: 5 }, defaultProps.config), '12345', '1234', '最短需输入5个字符。')
validateTest('文本框最大长度校验-中日韩字符', Object.assign({ maxLength: 5, cjkLength: 2 }, defaultProps.config), '中中', '中中中', '最长可输入5个字符。')
validateTest('文本框最小长度校验-中日韩字符', Object.assign({ minLength: 5, cjkLength: 2 }, defaultProps.config), '中中中', '中中', '最短需输入5个字符。')
validateTest('文本框正则校验-默认提示', Object.assign({ regExp: { expression: '^[0-9]$' } }, defaultProps.config), '0', 'a', '格式错误')
validateTest('文本框正则校验-自定提示', Object.assign({ regExp: { expression: '^[0-9]$', message: '提示' } }, defaultProps.config), '0', 'a', '提示')

test('文本框渲染', async () => {
  const component = renderer.create(
    <TextField {...defaultProps} ref={async (ref: any) => { }} />
  )

  const o = component.toJSON()
  expect(o).toMatchSnapshot()
})

test('文本框onChange', () => {
  return new Promise((resolve) => {
    const component = renderer.create(
      <TextField
        {...defaultProps}
        ref={async (ref: any) => { }}
        onChange={async (value) => {
          expect(value).toEqual('onChange')
          cleanup()
          resolve(true)
        }}
      />
    )

    const o = component.toJSON()
    if (o) {
      o[1].children[0].props.onClick()
    }
  })
})
