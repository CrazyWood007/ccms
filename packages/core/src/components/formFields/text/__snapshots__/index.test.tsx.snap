// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`文本框渲染 1`] = `
Array [
  "您当前使用的UI版本没有实现TextField组件。",
  <div
    style={
      Object {
        "display": "none",
      }
    }
  >
    <button
      onClick={[Function]}
    >
      onChange
    </button>
  </div>,
]
`;
